### PR TITLE
Exclude reading _pos column if it's not in the scan list

### PR DESCRIPTION
**Pull request by huaxingao**
In Spark batch reading, Iceberg reads additional columns when there are delete files. For instance, if we have a table
`test (int id, string data)` and a query `SELECT id FROM test`, the requested schema only contains the column `id`. However, to determine which rows are deleted (there is a `rowIdMapping` for this purpose), Iceberg appends `_pos` to the requested schema for position deletes, and append the equality filter column for equality deletes (suppose the equality delete is on column data). As a result, Iceberg will have `ColumnarBatchReader`s for these extra columns. In the case of position deletes, we actually don't need to read `_pos` to compute the `rowIdMapping`, so this PR excludes the `_pos` columns when building the `ColumnarBatchReader`. For equality deletes, while we need to read the equality filter column to compute the `rowIdMapping`, once we have the `rowIdMapping`, we should exclude the values of these extra columns from the `ColumnarBatch`. I will have a separate PR to fix equality delete.

In summary:
```
SELECT id FROM test
```
For position delete, the vectorized reader currently returns a `ColumnarBatch` that contains arrow vectors for both `id` and `_pos`. This PR will make iceberg not read `_pos` column, so the returned `ColumnarBatch` only contains an arrow vector for `id` only.

For equality delete (suppose the filter is on `data` column, the vectorized reader currently returns a `ColumnarBatch` that contains arrow vectors for both `id` and `data`. The goal is to return a `ColumnarBatch` that contains an arrow vector for `id` only. I will have a separate PR for this.

### CODE REVIEW

**Comment by huaxingao** [comment]
cc @szehon-ho @pvary @viirya

**Comment by huaxingao** [comment]
also cc @flyrain

**Comment by szehon-ho** [comment]
@huaxingao its a good find, im just wondering, where do we add _pos to the schema?  Can we just not do it there?  Just curious if its possible

**Comment by dramaticlly** [comment]
> @huaxingao its a good find, im just wondering, where do we add _pos to the schema? Can we just not do it there? Just curious if its possible

I think it might be from here https://github.com/apache/iceberg/blob/684f7a767c2c216a402b60b73d2d55ef605921a0/data/src/main/java/org/apache/iceberg/data/DeleteFilter.java#L260-L263

**Comment by huaxingao** [comment]
@szehon-ho I think we still need the `_pos` in the `requiredSchema` to build [`posAccessor`](https://github.com/apache/iceberg/blob/main/data/src/main/java/org/apache/iceberg/data/DeleteFilter.java#L97).

**Comment by huaxingao** [comment]
@pvary Thank you for your suggestion! You're correct that adding such a test would help prevent future changes from inadvertently affecting this behavior without notice. Currently, Spark doesn't check the schema when processing batch data, which is why an extra Arrow vector in `ColumnarBatch` doesn't cause error. However, Comet allocates arrays in a pre-allocated list and relies on the requested schema to determine how many columns are in the batch. If extra columns are returned to Comet, it will fail. While we currently don't have a test that fails due to extra columns, the integration of Comet will change this. Once Comet is integrated, the tests involving the Comet reader will fail if extra columns are present. I believe these Comet reader tests will serve as the tests you've suggested we add.

File: spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/BaseBatchReader.java [hunk 1]
```diff
@@ -125,4 +129,28 @@ private CloseableIterable<ColumnarBatch> newOrcIterable(
...
+  private Schema vectorizationSchema(SparkDeleteFilter deleteFilter) {
+    // For pos delete, deleteFilter has appended _pos to the required schema.
+    // For example, SELECT id, data FROM test, the requested schema is id and data. If there
+    // is position delete, deleteFilter will append _pos to the schema so the schema becomes
+    // id, data and _pos. However, vectorization reader only needs to read the requested columns,
+    // i.e. id and data, so we want to remove the _pos from the schema when building the
+    // vectorization reader. Before removing _pos, we need to make sure _pos is not explicitly
+    // selected in the query.
+    if (deleteFilter != null) {
+      if (deleteFilter.hasPosDeletes() && expectedSchema().findType("_pos") == null) {
```
Comment (viirya) [inline spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/BaseBatchReader.java:142]: Do we have a const for `_pos`?

Comment (huaxingao) [reply spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/BaseBatchReader.java]: Yes, we do. Changed to `MetadataColumns.ROW_POSITION.name()`

File: spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/BaseBatchReader.java [hunk 2]
```diff
@@ -81,14 +84,15 @@ private CloseableIterable<ColumnarBatch> newParquetIterable(
...
+    Schema vectorizationSchema = vectorizationSchema(deleteFilter);
 
     return Parquet.read(inputFile)
         .project(requiredSchema)
         .split(start, length)
         .createBatchedReaderFunc(
             fileSchema ->
                 VectorizedSparkParquetReaders.buildReader(
-                    requiredSchema, fileSchema, idToConstant, deleteFilter))
+                    vectorizationSchema, fileSchema, idToConstant, deleteFilter))
```
Comment (viirya) [inline spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/BaseBatchReader.java:95]: Is it just `expectedSchema`?

Comment (huaxingao) [reply spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/BaseBatchReader.java:95]: Not exactly. 
If no deletes, it's `expectedSchema`. 
If it's equality delete, it's `deleteFilter.requiredSchema()`, because it could be `expectedSchema` + equality filter column. For example:
```
SELECT id FROM table
```
supposed the equality delete has data == 'aaa'
then we do need to read the `data` column too, so it's `deleteFilter.requiredSchema()`, which is `id` + `data`

Comment (huaxingao) [reply spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/BaseBatchReader.java:95]: In the case of pos delete, I can't use `expectedSchema` either, because it could be both pos delete and equality delete.

**Comment by szehon-ho** [comment]
Sorry I still wanted to see if it can be done earlier, what do you think https://github.com/apache/iceberg/blob/main/spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/BatchDataReader.java#L99. this is for vectorized path right?  Can we pass in a flag to SparkDeleteFilter to not add the column?

Also the current fix is specific to Parquet, how about ORC?

**Comment by huaxingao** [comment]
@szehon-ho Thanks for the comment.

We actually also use the [requiredSchema](https://github.com/apache/iceberg/blob/fda2b3a5706fd580b0371e8a7c4b31d536eac0a3/spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/BaseBatchReader.java#L90), that's the schema with the `_pos` column. In ReadConf#[generateOffsetToStartPos](https://github.com/apache/iceberg/blob/main/parquet/src/main/java/org/apache/iceberg/parquet/ReadConf.java#L185), we actually need to know if pos delete exists. 

We can pass in a flag to `SparkDeleteFilter` to not add the `_pos` column, but then I think we need to add another flag to pass the hasPosDelete info to Parquet `ReaderBuilder`, and then pass to `ReadConfig`. 

ORC uses [expectedSchema()](https://github.com/apache/iceberg/blob/fda2b3a5706fd580b0371e8a7c4b31d536eac0a3/spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/BaseBatchReader.java#L125), the schema without _pos column, to build vectorized readers.

**Comment by szehon-ho** [comment]
Hm then can we we just add _pos to requiredSchema (with a comment)?  

Probably cleaner with a flag to ReadConf but not sure if its feasible.  fyi @aokolnychyi

File: spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/BatchDataReader.java [hunk 3]
```diff
@@ -93,10 +93,10 @@ protected CloseableIterator<ColumnarBatch> open(FileScanTask task) {
...
     Preconditions.checkNotNull(inputFile, "Could not find InputFile associated with FileScanTask");
 
-    SparkDeleteFilter deleteFilter =
-        task.deletes().isEmpty()
-            ? null
-            : new SparkDeleteFilter(filePath, task.deletes(), counter());
+    SparkDeleteFilter sparkDeleteFilter =
+        new SparkDeleteFilter(filePath, task.deletes(), counter(), true);
+
+    SparkDeleteFilter deleteFilter = task.deletes().isEmpty() ? null : sparkDeleteFilter;
```
Comment (flyrain) [inline spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/BatchDataReader.java:99]: We don't need delete filter if `task.deletes().isEmpty()`, but the new code always creates a filter object.
```
filter = null;
if(!task.deletes().isEmpty()) {
 filter = new filter()
}
```

File: data/src/main/java/org/apache/iceberg/data/DeleteFilter.java [hunk 4]
```diff
@@ -69,7 +69,8 @@ protected DeleteFilter(
       List<DeleteFile> deletes,
       Schema tableSchema,
       Schema requestedSchema,
-      DeleteCounter counter) {
+      DeleteCounter counter,
+      boolean isBatchReading) {
```
Comment (flyrain) [inline data/src/main/java/org/apache/iceberg/data/DeleteFilter.java:73]: Does batch reading necessarily mean no `_pos`? I believe not, as you can explicitly project it, like `select _pos from t`. We should give an accurate name, something like `needRowPosition` or `needRowPosCol`.

Comment (flyrain) [reply data/src/main/java/org/apache/iceberg/data/DeleteFilter.java:73]: Think a bit more, I'm considering if only the non-vector readers actually need an implicit `_pos` column. If that’s the case, would it make more sense to adjust this within RowReader by adding `_pos` there? This approach could simplify things by eliminating the need to check whether a reader is vectorized, especially since vectorization isn’t necessarily strongly correlated with the requirement for `_pos`.
Here is pseudo code to add it in class `RowDataReader`
```
    LOG.debug("Opening data file {}", filePath);
    expectedSchema().add(`_pos`);    <--- add it here
    SparkDeleteFilter deleteFilter =
        new SparkDeleteFilter(filePath, task.deletes(), counter(), false);
```

File: data/src/main/java/org/apache/iceberg/data/DeleteFilter.java [hunk 5]
```diff
@@ -93,7 +94,8 @@ protected DeleteFilter(
 
     this.posDeletes = posDeleteBuilder.build();
     this.eqDeletes = eqDeleteBuilder.build();
-    this.requiredSchema = fileProjection(tableSchema, requestedSchema, posDeletes, eqDeletes);
+    this.requiredSchema =
+        fileProjection(tableSchema, requestedSchema, posDeletes, eqDeletes, isBatchReading);
```
Comment (flyrain) [inline data/src/main/java/org/apache/iceberg/data/DeleteFilter.java:98]: One question I asked myself is whether it impacts the metadata column read? It seems not, but the method `DeleteFilter::fileProjection` seems a bit hard to read, we can refactor it later. It makes more sense to make it a static until method instead of instance method. It's a bit weird to pass the schema to delete filter, then get it back from the filter. This seems something we can improve on it as a follow up.

File: parquet/src/main/java/org/apache/iceberg/parquet/ReadConf.java [hunk 6]
```diff
@@ -181,8 +184,8 @@ boolean[] shouldSkip() {
     return shouldSkip;
   }
 
-  private Map<Long, Long> generateOffsetToStartPos(Schema schema) {
-    if (schema.findField(MetadataColumns.ROW_POSITION.fieldId()) == null) {
+  private Map<Long, Long> generateOffsetToStartPos() {
+    if (hasPositionDelete) {
```
Comment (flyrain) [inline parquet/src/main/java/org/apache/iceberg/parquet/ReadConf.java:188]: I'd recommend to do it in the caller like this 
```
    Map<Long, Long> offsetToStartPos = hasPositionDelete? generateOffsetToStartPos(): null;
```

File: spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/BaseRowReader.java [hunk 7]
```diff
@@ -98,6 +98,7 @@ private CloseableIterable<InternalRow> newParquetIterable(
         .filter(residual)
         .caseSensitive(caseSensitive())
         .withNameMapping(nameMapping())
+        .hasPositionDelete(readSchema.findField(MetadataColumns.ROW_POSITION.fieldId()) == null)
```
Comment (flyrain) [inline spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/BaseRowReader.java:101]: Do we need it for row reader? Can we use a default value here?

File: spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/BaseBatchReader.java [hunk 8]
```diff
@@ -97,6 +98,7 @@ private CloseableIterable<ColumnarBatch> newParquetIterable(
         // read performance as every batch read doesn't have to pay the cost of allocating memory.
         .reuseContainers()
         .withNameMapping(nameMapping())
+        .hasPositionDelete(hasPositionDelete)
```
Comment (flyrain) [inline spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/BaseBatchReader.java:101]: Wondering if `withPositionDelete` is more suitable

Comment (flyrain) [reply spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/BaseBatchReader.java:101]: The key here is determining whether we want to compute row offsets specifically for the filtered row groups. This decision doesn’t have to be directly tied to the presence of position deletes. Perhaps a name like `needRowGroupOffset` would better capture this intent and improve clarity.

**Review by flyrain — COMMENTED** [review]
Thanks @huaxingao for working on it. Sorry for the delay. Left some comments.

**Comment by huaxingao** [comment]
Thanks a lot @flyrain and @szehon-ho for your review! I've thought this over: I feel the [original change](https://github.com/apache/iceberg/pull/11390/commits/9b4e56d7dd9a6571655fd966ba29601d03a787de)  is much simpler. If the original one looks OK to you, I will revert to it. Thanks a lot!

File: parquet/src/main/java/org/apache/iceberg/parquet/ReadConf.java (hunk 6 above)
Comment (flyrain) [reply parquet/src/main/java/org/apache/iceberg/parquet/ReadConf.java:188]: BTW, we don't actually need this if this PR#10107 is in, which removes related complexity. The parquet side is ready  for a while. We may get #10107 in.

File: data/src/main/java/org/apache/iceberg/data/DeleteFilter.java [hunk 9]
```diff
@@ -251,13 +253,14 @@ private static Schema fileProjection(
...
-      List<DeleteFile> eqDeletes) {
+      List<DeleteFile> eqDeletes,
+      boolean needRowPosCol) {
     if (posDeletes.isEmpty() && eqDeletes.isEmpty()) {
       return requestedSchema;
     }
 
     Set<Integer> requiredIds = Sets.newLinkedHashSet();
-    if (!posDeletes.isEmpty()) {
+    if (!posDeletes.isEmpty() && needRowPosCol) {
```
Comment (sfc-gh-ygu) [inline data/src/main/java/org/apache/iceberg/data/DeleteFilter.java:263]: Nit: switch two condition to make it more readable. 
```
if (needRowPosCol && !posDeletes.isEmpty()) {}
```

**Review by flyrain — APPROVED** [review]
+1 Pending pipeline.

**Review by szehon-ho — APPROVED** [review]
Looks good to me

**Comment by szehon-ho** [comment]
Merged, thanks @huaxingao , and also @flyrain for review, and @pvary @dramaticlly @viirya for other reviews

**Comment by huaxingao** [comment]
Thanks a lot! @flyrain @szehon-ho @viirya @dramaticlly @pvary
